### PR TITLE
[Ready to Merge] Improved docs for classify_sysode

### PR DESCRIPTION
**Pull request by toolforger**

### CODE REVIEW

**Comment by toolforger** [comment]
The original description was confusing, at least to me (who knows what an ODE is but does not know much beyond that).
The original docstring was by @KDMOE (Kundan Kumar), who was mentored by @tjl (Tim Lahey) and @flacjacket (Sean Vig), so if any of these is around, please comment and correct any omissions and errors.

Things that I am not happy about myself:
1) It seems that "coefficient" could mean constants as well as arbitrary expressions. I am not sure that that is standard mathematical terminology.
2) With "order of derivation applied to f(x)", I mean 1 for first derivation, 2 for second derivation, etc. It would be  nice if a native math-English speaker could fill in the proper terminology.
3) The docstring says "function" in several places. I would like to document the class that could rightfully occur there - UndefinedFunction? One of its superclasses?

**Comment by tjl** [comment]
Overall, I prefer the original text. While I admit it could be cleaned up, I find this edit to be terse.

Function is just a Sympy function. There's no real need to clarify that further. Plus, there's no need to refer to them as "function objects" in the docstring, just function is fine.

**Comment by toolforger** [comment]
Well, at least I'd uppercase "Function" so people are alerted that we're talking about SymPy objects, not about Python functions.
Good to know that it's arbitrary Functions, I'll update it.

To be frank, the original text is horrible:
- It's talking about "first" and "second" entries in dicts, which is not just inaccurate but wrong. Python dicts do not have an order, each implementation may return the entries in a different, albeit consistent order. In fact pypy deviates from CPython, and I suspect this misunderstanding about dict orderedness is causing test failures.
- It's identifying some dict entries by name ('func' etc.), some by number ('fourth entry').
- The last paragraph doesn't make any sense at all.

I tend to start with maximum terseness (it's easiest to check for correctness), then add detail as needed, so any proposals welcome.
I do think I have all the information of the original docstring. If anything is missing, please tell me.

**Comment by tjl** [comment]
I'll see what I can do. I'm fine with it being terse, but I think a bit more detail is necessary. 

Oh, and it's derivatives, not derivations.

Make your edits and I'll try and spend some time on adding detail where I think is necessary.

**Comment by toolforger** [comment]
I have waited a bit for a fresher look at the docstring.
I hope it has improved.

My understanding of an ODE is that each equation if a sum-of-products, each product consisting of a "bare function" (which appear as a derivative somewhere in the ODE, and which are what we're solving the ODE for), and some other expression that contains none of the functions that we solve the ODE for (the "coefficient").
I hope this is a largely faithful description of what an ODE is; if not, the docstring is most likely subtly or openly wrong in places.

File: sympy/solvers/ode.py [hunk 1]
```diff
@@ -1346,8 +1349,8 @@ def _sympify(eq):
     matching_hints = {"no_of_equation":i+1}
     matching_hints['eq'] = eq
     if i==0:
-        raise ValueError("classify_sysode() woks for systems of ODEs. For"
-        " single ODE equation solving classify_ode should be used")
+        raise ValueError("classify_sysode() works for systems of ODEs. For"
+        " single equation ODE solving, classify_ode should be used")
```
Comment (cbm755) [inline sympy/solvers/ode.py]: Suggest: "For scalar ODEs, classifiy_ode..."

File: sympy/solvers/ode.py [hunk 2]
```diff
@@ -1278,31 +1278,34 @@ def _test_term(coeff, order):
...
-    is same as is given references.
+    Returns a dictionary of parameter names and values that define the system
+    of ordinary differential equations in ``eq``.
+    The parameters are further used in
+    :py:meth:`~sympy.solvers.ode.dsolve` for solving that system.
+
+    The parameter names and values are:
+
+    'is_linear' (boolean), which tells whether the given system is linear (i.e.
+    whether all coefficients are constant).
```
Comment (cbm755) [inline sympy/solvers/ode.py]: linear != constant coefficient.

Comment (toolforger) [inline sympy/solvers/ode.py]: Oh, incomplete math background on my side then.
Fixed.

**Comment by toolforger** [comment]
@cbm755 ping?

**Comment by toolforger** [comment]
@tjl ping?

File: sympy/solvers/ode.py [hunk 3]
```diff
@@ -1278,31 +1278,34 @@ def _test_term(coeff, order):
...
+
+    The parameter names and values are:
+
+    'is_linear' (boolean), which tells whether the given system is linear.
+
+    'func' (list) contains the :py:class:`~sympy.core.function.Function`s that
+    appear with a derivative in the ODE, i.e. those that we are trying to solve
+    the ODE for.
+
+    'order' (dict) with the maximum derivative for each element of the 'func'
```
Comment (cbm755) [inline sympy/solvers/ode.py]: is it really a dict?

Comment (toolforger) [inline sympy/solvers/ode.py]: Yes. It's a list in `dsolve` (which might be wrong actually), but it is a dict here.
It's doing `order = dict` below, and `matching_hints['order'] = order` to set the result.

Comment (cbm755) [inline sympy/solvers/ode.py]: ok, I see it gets replaced later before the solvers get it.

**Comment by cbm755** [comment]
I'm +1 assuming @tjl is ok with it.

**Comment by toolforger** [comment]
@tjl ping?

**Comment by tjl** [comment]
Sorry for the delay, but I've been ill. I'm good. This can be merged.

**Comment by toolforger** [comment]
Oh. Best wishes for getting well again.
Thanks :-)

**Comment by flacjacket** [comment]
@toolforger This can't be cleanly merged, if you can fix that we can merge this in

**Comment by toolforger** [comment]
Ah, I didn't notice that it fell out of mergeability.
Fixed. I hope I didn't goof up during squashing.

**Comment by flacjacket** [comment]
Looks good, thanks

**Pull request merged**